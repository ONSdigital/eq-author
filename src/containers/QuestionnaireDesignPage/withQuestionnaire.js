import getQuestionnaireQuery from "schema/getQuestionnaire.graphql";
import { graphql } from "react-apollo";

import { find, get } from "lodash";

const findById = (collection, id) => find(collection, { id: parseInt(id, 10) });

export const mapResultsToProps = ({ data, ownProps }) => {
  const { questionnaire, loading } = data;
  const { sectionId, pageId } = ownProps;

  const props = {
    questionnaire,
    loading
  };

  if (questionnaire) {
    props.section = findById(questionnaire.sections, sectionId);
    props.page = findById(props.section.pages, pageId);
    props.answers = get(props.page, "answers", []);
  }

  return props;
};

export default graphql(getQuestionnaireQuery, {
  props: mapResultsToProps,
  options: props => ({ variables: { id: props.questionnaireId } })
});
