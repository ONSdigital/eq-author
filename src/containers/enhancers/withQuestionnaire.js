import getQuestionnaireQuery from "graphql/getQuestionnaire.graphql";
import { graphql } from "react-apollo";
import { get, pick, flow } from "lodash/fp";

export const mapResultsToProps = flow(
  get("data"),
  pick(["loading", "questionnaire"])
);

export const mapPropToOptions = props => ({
  variables: { id: props.questionnaireId }
});

export default graphql(getQuestionnaireQuery, {
  props: mapResultsToProps,
  options: mapPropToOptions
});
