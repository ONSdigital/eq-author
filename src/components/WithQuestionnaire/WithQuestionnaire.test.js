import React from "react";

import withQuestionnaire from "components/WithQuestionnaire";
import { shallow } from "enzyme";

let wrapper;

let handleUpdate;

let WrappedComponent;

describe("components/WithQuestionnaire", () => {
  const Child = props => <div {...props} />;
  Child.displayName = "Child";

  beforeEach(() => {
    handleUpdate = jest.fn();

    WrappedComponent = withQuestionnaire(Child);

    wrapper = shallow(
      <WrappedComponent onUpdate={handleUpdate} loading={false} />
    );
  });

  it("should not render when loading", () => {
    wrapper.setProps({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it("should store questionnaire props as state", () => {
    const questionnaire = {
      title: "My Title"
    };
    wrapper.setProps({ questionnaire });
    expect(wrapper.state().questionnaire).toEqual(questionnaire);
  });

  it("should set the displayName", () => {
    expect(WrappedComponent.displayName).toEqual("WithQuestionnaire(Child)");
  });

  it("should catch and prevent form submit events", () => {
    const preventDefault = jest.fn();
    wrapper.simulate("submit", { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  it("should store updated values in state", () => {
    wrapper.simulate("change", {
      name: "questionnaire.title",
      value: "My Title"
    });

    expect(wrapper.state().questionnaire).toEqual({ title: "My Title" });
  });

  it("should save to API on blur event", () => {
    wrapper
      .simulate("change", {
        name: "questionnaire.title",
        value: "My Title"
      })
      .simulate("blur");
    expect(handleUpdate).toHaveBeenCalledWith(
      expect.objectContaining({ questionnaire: { title: "My Title" } })
    );
  });
});
