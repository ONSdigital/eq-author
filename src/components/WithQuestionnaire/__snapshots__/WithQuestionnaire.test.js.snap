// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/WithQuestionnaire should not render when loading 1`] = `null`;
